src/mpid/ch4/netmod/ofi/ofi_win.h

```diff
@@ -998,66 +998,76 @@ static inline int MPIDI_NM_mpi_win_allocate_shared(MPI_Aint size,
         MPIR_ERR_SETANDSTMT(mpi_errno, MPI_ERR_NO_MEM, goto fn_fail, "**nomem");
     }
 
-    if (comm_ptr->rank == 0) {
-        int map_flags = MAP_SHARED;
+    int iter = MPIR_CVAR_CH4_SHM_SYMHEAP_RETRY;
+    unsigned anyfail = 1;
 
-        map_ptr = MPIDI_CH4R_generate_random_addr(mapsize);
+    while (anyfail && --iter > 0) {
+        if (comm_ptr->rank == 0) {
+            int map_flags = MAP_SHARED;
+
+            map_ptr = MPIDI_CH4R_generate_random_addr(mapsize);
 #ifdef USE_SYM_HEAP
-        if (!MPIDI_CH4R_check_mapaddr(map_ptr))
-            map_flags |= MAP_FIXED;     /* Set fixed only when generated a valid address.
+            if (!MPIDI_CH4R_check_mapaddr(map_ptr))
+                map_flags |= MAP_FIXED; /* Set fixed only when generated a valid address.
                                          * Otherwise we allow system to pick up one. */
 #endif
-        map_ptr = mmap(map_ptr, mapsize, PROT_READ | PROT_WRITE, map_flags, fd, 0);
+            map_ptr = mmap(map_ptr, mapsize, PROT_READ | PROT_WRITE, map_flags, fd, 0);
 
-        if (map_ptr == NULL || map_ptr == MAP_FAILED) {
-            close(fd);
+            if (map_ptr == NULL || map_ptr == MAP_FAILED) {
+                close(fd);
 
-            if (first)
-                shm_unlink(shm_key);
+                if (first)
+                    shm_unlink(shm_key);
 
-            MPIR_ERR_SETANDSTMT(mpi_errno, MPI_ERR_NO_MEM, goto fn_fail, "**nomem");
+                MPIR_ERR_SETANDSTMT(mpi_errno, MPI_ERR_NO_MEM, goto fn_fail, "**nomem");
+            }
         }
 
         mpi_errno = MPIR_Bcast_impl(&map_ptr, 1, MPI_UNSIGNED_LONG, 0, comm_ptr, &errflag);
 
         if (mpi_errno != MPI_SUCCESS)
             goto fn_fail;
 
-        MPIDI_CH4U_WIN(win, mmap_addr) = map_ptr;
-        MPIDI_CH4U_WIN(win, mmap_sz) = mapsize;
-    }
-    else {
-        mpi_errno = MPIR_Bcast_impl(&map_ptr, 1, MPI_UNSIGNED_LONG, 0, comm_ptr, &errflag);
-
-        if (mpi_errno != MPI_SUCCESS)
-            goto fn_fail;
-
-        int map_flags = MAP_SHARED;
+        unsigned test = 0;
```

Comment:
I think this variable name should be more self-explanatory.
Also, it looks to me that it has two different usages in the code. Probably you should introduce two variables like `mmap_needed` and `failure` or something, if I understood your intention right.